apps/src/components/map-layers/s2d-variable-values.tsx

```diff
@@ -30,250 +81,799 @@ const tooltipTemperatureRange = __(
+
+	// If the percentage/outcome is bigger than the highest value
+	if (colourIndex === -1) {
+		return defaultColor;
+	}
+
+	// If the next colour is not in the same outcome
+	if (!isSameThousand(colorMap.quantities[colourIndex], queryQuantity)) {
+		return defaultColor;
+	}
+
+	let colour = colours[colourIndex];
+
+	// If the colour is white, we find the next none-white colour
+	if (isWhite(colour)) {
+		// First, set the colour to a default "grey", in case we don't find
+		// another replacement colour.
+		colour = defaultColor;
+
+		for (let i = colourIndex + 1; i < colours.length; i++) {
+			// Stop if we are now in the next outcome's colours
+			if (!isSameThousand(colorMap.quantities[i], queryQuantity)) {
+				break;
+			}
+
+			const nextColour = colours[i];
+			if (!isWhite(nextColour)) {
+				colour = nextColour;
+				break;
+			}
+		}
+	}
+
+	return colour;
+};
+
+/**
+ * Main component for the location popup of a S2D variable. Wrapper containing
+ * the logic, around the actual component.
+ */
+export const S2DVariableValues = ({ latlng }: S2DVariableValuesProps) => {
+	const { releaseDate } = useS2D();
+	const { climateVariable } = useClimateVariable();
+
+	// If we have everything we need to start loading the location's data
+	const readyToLoad = !!(releaseDate && climateVariable);
+	const [locationData, setLocationData] =
+		useState<LocationS2DData | null>(null);
+	const lastLoadingRef = useRef<string | null>(null);
+
+	const hasDataLoaded = !!(locationData && releaseDate);
+
+	const variableId = climateVariable?.getId() ?? '';
+	const variableName = climateVariable?.getTitle() ?? '';
+	const unit = climateVariable?.getUnit() ?? '';
+	const frequency = (climateVariable?.getFrequency() ??
+		FrequencyType.MONTHLY) as S2DFrequencyType;
+	const forecastType =
+		climateVariable?.getForecastType() ?? ForecastTypes.UNUSUAL;
+	const forecastDisplay =
+		climateVariable?.getForecastDisplay() ?? ForecastDisplays.FORECAST;
+	const dateRange = climateVariable?.getDateRange();
+	const dateRangeStart = dateRange ? dateRange[0] : null;
+
+	useEffect(() => {
+		const loadingKey = `${latlng.lat}-${latlng.lng}-${variableId}-${frequency}-${dateRangeStart}`;
+		const period = dateRangeStart ? utc(dateRangeStart) : null;
+
+		if (!readyToLoad || !period || lastLoadingRef.current === loadingKey) {
+			return;
+		}
+
+		const abortController = new AbortController();
+
+		const fetchData = async () => {
+			const loadedLocationData = await fetchS2DLocationData(
+				latlng,
+				variableId,
+				frequency,
+				period,
+				{ signal: abortController.signal }
+			);
+
+			if (abortController.signal.aborted) {
+				return;
+			}
+
+			lastLoadingRef.current = null;
+			setLocationData(loadedLocationData);
+		};
+
+		lastLoadingRef.current = loadingKey;
+		setLocationData(null);
+		fetchData();
+
+		return () => {
+			lastLoadingRef.current = null;
+			abortController.abort();
+		};
```

Comment:
I had been thinking about this block and was pondering why we *fire and forget* `fetchData`. Yes the current `fetchS2DLocationData` is a mock and not an actual HTTP call -- I'm talking about when this pattern and actual HTTP.

I didn't realize that react would call the cleanup function.

So, I've removed the comments I had here about that.